### ISSUE TITLE
IsRFC* improvements

### ISSUE THREAD

**Issue by MithrilMan** [MetacoSA/NBitcoin#793]
https://github.com/MetacoSA/NBitcoin/blob/a615d201aff38b55e954fa3a184a0a439f202b12/NBitcoin/IpExtensions.cs#L123-L127

this code is awkward (`bytes[15 - 15]` ==> `bytes[0]`, probably the result of a typo long time ago?
compare it to bitcoin one
https://github.com/bitcoin/bitcoin/blob/47d981e8273804a040d71665a4cb16038d6717e1/src/netaddress.cpp#L191-L194

bitcoin uses there a GetByte method probably to deal with endianness but ended up being just a
```
unsigned int CNetAddr::GetByte(int n) const
{
    return ip[15-n];
}
```
since you didn't implemented that call, why don't remove all these 15-x from code?

another think I'd want to point out is 
```
public static bool IsRFC4862(this IPAddress address)
{
	var bytes = address.GetAddressBytes();
	byte[] pchRFC4862 = new byte[] { 0xFE, 0x80, 0, 0, 0, 0, 0, 0 };
	return ((Utils.ArrayEqual(bytes, 0, pchRFC4862, 0, pchRFC4862.Length) ? 0 : 1) == 0);
}
```

everytime IsRFC4862 is called there is a new allocation of pchRFC4862, I'd suggest to put it as static

using Span it may seems like this (not tested)
```
private static readonly byte[] pchRFC4862 = new byte[] { 0xFE, 0x80, 0, 0, 0, 0, 0, 0 };
public static bool IsRFC4862(this byte[] bytes) {
   Span<byte> span = bytes.Length > pchRFC4862.Length ? bytes.AsSpan(0, pchRFC4862.Length) : bytes;
   return span.SequenceEqual(pchRFC4862);
}
```

**Comment by MithrilMan** [comment]
another improvement could be having these extensions accepting a byte array, because IsRoutable extension actually causes to generate `address.GetAddressBytes()` multiple times (I've to check dotnet source code to see if it means new allocation everytime or not)

**Comment by NicolasDorier** [comment]
I agree, yes I monkey ported it from bitcoin core. I would appreciate a PR if you have time.

**Comment by MithrilMan** [comment]
Actually I'm quite busy but leave this issue open, when I'll find time I'll create a PR about this
